
# RNA-seq Pipeline for _Mycobacterium tuberculosis_

A basic RNA-sequencing pipeline for QC, alignment, and counting of paired-end reads generated by Illumina next gen sequencing
for the Baughn Lab. This pipeline in its current iteration is meant to run through submission to the MSI Slurm queue as a single file and load all modules required directly from MSI. 

## Pipeline Summary

![RNAseq Analysis Pipeline Updated](https://user-images.githubusercontent.com/115341984/196510238-d3390d45-9cd4-4e87-9fdb-0b84671a5d10.png)

    1. Quality Check (FastQc)
    2. T-overhang trimming (Cutadapt)
    3. rRNA contamination removal (ribodetector)
    3. Alignment (STAR)
    4. Counting (FeatureCounts)
    5. QC (RSeQC)
    6. Combine all outputs (MultiQC)

## Usage

### Running the pipeline

```shell
qsub RNAseq_Analysis.sh -f [/path/to/RNAseqfolder] -i [/path/to/inputfiledirectory] -d [/path/to/DNAreference] -a [/path/to/Annotationreference]
```




## Dependencies
TBAdded


## To Add
- Replace Cutadapt with package that runs faster and is more dynamic (Trimmomatic/Trim Galore!)

More extensive quality control with:

- BEDtools, bedGraphToBigWig
- Qualimap


